controllers/topology/internal/contract/controlplane_test.go

```diff
@@ -94,3 +150,156 @@ func TestControlPlane(t *testing.T) {
 		g.Expect(got).To(Equal(metadata))
 	})
 }
+
+func TestControlPlaneIsUpgrading(t *testing.T) {
+	tests := []struct {
+		name          string
+		obj           *unstructured.Unstructured
+		wantUpgrading bool
+	}{
+		{
+			name: "should return false for stable control plane",
+			obj: &unstructured.Unstructured{Object: map[string]interface{}{
+				"spec": map[string]interface{}{
+					"version": "1.2.3",
+				},
+				"status": map[string]interface{}{
+					"version": "1.2.3",
+				},
+			}},
+			wantUpgrading: false,
+		},
+		{
+			name: "should return false if status is not set on control plane",
+			obj: &unstructured.Unstructured{Object: map[string]interface{}{
+				"spec": map[string]interface{}{
+					"version": "1.2.3",
+				},
+			}},
+			wantUpgrading: false,
+		},
+		{
+			name: "should return false if status.version is not set on control plane",
+			obj: &unstructured.Unstructured{Object: map[string]interface{}{
+				"spec": map[string]interface{}{
+					"version": "1.2.3",
+				},
+				"status": map[string]interface{}{},
+			}},
+			wantUpgrading: false,
+		},
+		{
+			name: "should return true if spec and status versions do not match",
+			obj: &unstructured.Unstructured{Object: map[string]interface{}{
+				"spec": map[string]interface{}{
+					"version": "1.2.3",
+				},
+				"status": map[string]interface{}{
+					"version": "1.2.1",
+				},
+			}},
+			wantUpgrading: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			actual, _ := ControlPlane().IsUpgrading(tt.obj)
+			g.Expect(actual).To(Equal(tt.wantUpgrading))
+		})
+	}
+}
+
+func TestControlPlaneIsScaling(t *testing.T) {
+	tests := []struct {
+		name        string
+		obj         *unstructured.Unstructured
+		wantScaling bool
+	}{
+		{
+			name: "should return false for stable control plane",
+			obj: &unstructured.Unstructured{Object: map[string]interface{}{
+				"spec": map[string]interface{}{
+					"replicas": int64(2),
```

Comment:
I havent observed any problem when testing. I am going to say it is not a problem. I think there was also a discussion about making replicas an int64 fields. I am gonna leave this in for now.

Note: the contract package introduces replicas as int64 because we chose to switch to int64 in the desired_state.go. Ref: https://github.com/kubernetes-sigs/cluster-api/pull/5140/files#diff-0f241e843545fa7d707779d317c01ac24d679ea305401d22398ae07728ad454cL169